clap_complete/src/dynamic/completer.rs

```diff
@@ -344,3 +358,65 @@ fn subcommands(p: &clap::Command) -> Vec<(String, Option<StyledStr>)> {
         .flatten()
         .collect()
 }
+
+/// A completion candidate defination
+///
+/// This makes it easier to add more fields to completion candidate,
+/// rather than using `(OsString, Option<StyledStr>)` or `(String, Option<StyledStr>)` to represent a completion candidate
+pub struct CompletionCandidate {
+    /// Main completion candidate content
+    pub content: OsString,
+
+    /// Help message with a completion candidate
+    pub help: Option<StyledStr>,
+}
+
+impl From<(OsString, Option<StyledStr>)> for CompletionCandidate {
+    fn from(value: (OsString, Option<StyledStr>)) -> Self {
+        Self {
+            content: value.0,
+            help: value.1,
+        }
+    }
+}
+
+impl From<(String, Option<StyledStr>)> for CompletionCandidate {
+    fn from(value: (String, Option<StyledStr>)) -> Self {
+        Self {
+            content: OsString::from(value.0),
+            help: value.1,
+        }
+    }
+}
+
+impl Into<(OsString, Option<StyledStr>)> for CompletionCandidate {
+    fn into(self) -> (OsString, Option<StyledStr>) {
+        (self.content, self.help.clone())
+    }
+}
+
+impl PartialEq for CompletionCandidate {
+    fn eq(&self, other: &Self) -> bool {
+        self.content == other.content
+    }
+}
+
+impl Eq for CompletionCandidate {}
+
+impl PartialOrd for CompletionCandidate {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.content.cmp(&other.content))
+    }
+}
+
+impl Ord for CompletionCandidate {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.content.cmp(&other.content)
+    }
+}
```

Comment:
Specializing operators like this can be tricky and should be clearly documented that its done and why.  Most likely, it shouldn't be.